Bug Fixes

olamiral opened the pull request:
- URL Handler: parseUrl was not parsing URLs correctly when deleting
files recursively

- Resource Locator Implementation: paths and names were concatenated
without path separator in method resolveInContext

mbechler commented:
Can you describe under which circumstances these issues occur?

While this may be unexpected and quite inconvenient, jcifs has always required that URLs to directories are specified with a trailing slash by the user.

olamiral commented:
Sure!

Suppose you have a directory in smb://smb_host/share_root/parentDir/childDir.
The following will fail with "The system cannot find the file specified.":

`SmbFile parentDir = new SmbFile("smb://smb_host/share_root/parentDir", context);
parentDir.delete();`

I'm new to jcifs, and I was not aware of that! Thank you for clarifying it. I've updated my unit tests to take that in account and now they work without the workaround I've implemented. I wish I could know about it before.

Based on what you sair, one thing that caught my attention was the fact that running  `new SmbFile("smb://smb_host/share_root/someDir").isDirectory()`, jcifs returns true. If the trailing slash is required by jcifs when the user wants to execute any operation on a path that is a directory, then the above command should return false, because either the path would not exist or it would be a file.

Anyway, knowing how jcifs works a little better, this pull request can be rejected.

olamiral commented:
By the way, I've noticed that the modification done in SmbResourceLocatorImpl fixed the problem related with recursive directory deletion (my unit tests break without it, and works fine with that fix). 

In the other hand, the modification done in the Smb Url Handler was not necessary after I updated my unit tests appending a trailing slash when the paths are directories.

dumh commented:
Why isn't it merged yet?? These fixes are so needed! SmbFile.list() and .delete() don't work as expected because of this.

mbechler commented:
Please watch your tone. 

Note, that as I mentioned at least some of the described issue appears to be actually "documented" behavior (even though this is admittetly stupid). Changing that however I believe has some other implications which need further investigation, making this more complex than just merging the patch.

But, I have to admit I lost track of it, at least the recursive deletion issue sounds like a legtimate one. I'll try to look into it some time soon.

dumh commented:
I'm sorry for the double question marks, it was only to express my utter surprise for such a bug not being fixed yet, no offence meant really.. I understand and appreciate the hard work you do for the project, thank you!

I got your point about backward compatibility. Maybe introduce a way of selecting the behaviour, keeping the old one as a legacy.
